docs/architecture/adr-069-unordered-account.md

```diff
@@ -0,0 +1,177 @@
+
+### Transaction Format
+
+```protobuf
+message TxBody {
+  ...
+
+  boolean unordered = 4; 
+}
+```
+
+### `DedupTxHashManager`
+
+```golang
+// can reduce frequency we check the expiration.
+const ExpireCheckInterval = 1
+
+// DedupTxHashManager contains the tx hash dictionary for duplicates checking,
+// and expire them when block number progresses.
+type DedupTxHashManager struct {
+  // tx hash -> expire block number
+  // for duplicates checking and expiration
+  hashes map[TxHash]uint64
+}
+
+func (dtm *DedupTxHashManager) Contains(hash TxHash) (ok bool) {
+  dtm.mutex.RLock()
+  defer dtm.mutex.RUnlock()
+
+  _, ok = dtm.hashes[hash]
+  return
+}
+
+func (dtm *DedupTxHashManager) Size() int {
+  dtm.mutex.RLock()
+  defer dtm.mutex.RUnlock()
+
+  return len(dtm).hashes
+}
+
+func (dtm *DedupTxHashManager) Add(hash TxHash, expire uint64) (ok bool) {
+  dtm.mutex.Lock()
+  defer dtm.mutex.Unlock()
+
+  dtm.hashes[hash] = expire
+  return
+}
+
+// EndBlock remove expired tx hashes, need to wire in abci cycles.
+func (dtm *DedupTxHashManager) EndBlock(ctx sdk.Context) {
+  if ctx.BlockNumber() % ExpireCheckInterval != 0 {
+    return
+  }
+
+  dtm.mutex.Lock()
+  defer dtm.mutex.Unlock()
+
+  for k, expire := range dtm.hashes {
+    if ctx.BlockNumber() > expire {
+      delete(dtm.hashes, k)
+    }
+  }
+}
+```
+
+### Ante Handlers
+
+Bypass the nonce decorator for un-ordered transactions.
+
+```golang
+func (isd IncrementSequenceDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+  if tx.UnOrdered() {
+    return next(ctx, tx, simulate)
+  }
+  
+  // the previous logic
+}
+```
+
+A decorator for the new logic.
+
+```golang
+type TxHash [32]byte
+
+const (
+  // MaxNumberOfTxHash * 32 = 128M max memory usage
+  MaxNumberOfTxHash = 1024 * 1024 * 4
+
+  // MaxUnOrderedTTL defines the maximum ttl an un-order tx can set
+  MaxUnOrderedTTL = 1024
+)
+
+type DedupTxDecorator struct {
+  m *DedupTxHashManager
+}
+
+func (dtd *DedupTxDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+  // only apply to un-ordered transactions
+  if !tx.UnOrdered() {
+    return next(ctx, tx, simulate)
+  }
+
+  if tx.TimeoutHeight() == 0 {
+    return nil, errorsmod.Wrap(sdkerrors.ErrLogic, "unordered tx must set timeout-height")
+  }
+
+  if tx.TimeoutHeight() > ctx.BlockHeight() + MaxUnOrderedTTL {
+    return nil, errorsmod.Wrapf(sdkerrors.ErrLogic, "unordered tx ttl exceeds %d", MaxUnOrderedTTL)
+  }
+
+  if !ctx.IsCheckTx() {
+    // a new tx included in the block, add the hash to the dictionary
```

Comment:
For self reminding, check for duplicates in deliverTx as well, checktx is not reliable